sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,192 @@ def receive(self, windowed_value):
     else:
-      assert context is not None
-      self.context = context
+      return PerWindowInvoker(
+          signature, context, side_inputs, input_args, input_kwargs)
 
-    class ArgPlaceholder(object):
-      def __init__(self, placeholder):
-        self.placeholder = placeholder
+  def invoke_process(self, element, process_output_fn):
+    raise NotImplementedError
 
-    # Stash values for use in dofn_process.
-    self.side_inputs = side_inputs
-    self.has_windowed_inputs = not all(
-        si.is_globally_windowed() for si in self.side_inputs)
-
-    self.args = args if args else []
-    self.kwargs = kwargs if kwargs else {}
-    self.dofn = fn
-    self.dofn_process = fn.process
-
-    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
-    defaults = defaults if defaults else []
-    self_in_args = int(self.dofn.is_process_bounded())
-
-    self.use_simple_invoker = (
-        not side_inputs and not args and not kwargs and not defaults)
-    if self.use_simple_invoker:
-      # As we're using the simple invoker we don't need to compute placeholders
-      return
+  def invoke_start_bundle(self, process_output_fn):
+    defaults = self.signature.start_bundle_method.defaults
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.start_bundle_method.call(args, {}))
+
+  def invoke_finish_bundle(self, process_output_fn):
+    defaults = self.signature.finish_bundle_method.defaults
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.finish_bundle_method.call(args, {}))
 
-    self.has_windowed_inputs = (self.has_windowed_inputs or
-                                core.DoFn.WindowParam in defaults)
+
+class SimpleInvoker(DoFnInvoker):
+  """An invoker that processes elements ignoring windowing information."""
+
+  def invoke_process(self, element, process_output_fn):
+    process_output_fn(element, self.signature.process_method.call(
+        [element.value], {}))
+
+
+class PerWindowInvoker(DoFnInvoker):
+  """An invoker that processes elements considering windowing information."""
+
+  def __init__(self, signature, context, side_inputs, input_args, input_kwargs):
+    super(PerWindowInvoker, self).__init__(signature)
+    self.side_inputs = side_inputs
+    self.context = context
+    default_arg_values = signature.process_method.defaults
+    self.has_windowed_inputs = (
+        not all(si.is_globally_windowed() for si in side_inputs) or
+        (core.DoFn.WindowParam in default_arg_values))
 
     # Try to prepare all the arguments that can just be filled in
     # without any additional work. in the process function.
     # Also cache all the placeholders needed in the process function.
 
     # Fill in sideInputs if they are globally windowed
+
+    global_window = GlobalWindow()
+
+    args = input_args if input_args else []
```

Comment:
Thanks for changing to input_args. I think we can improve it a bit more as it is still confusing to understand what is input_args/args/final_args ??